File: Java/UsbAndPS.java
/*  https://codeforces.com/problemset/problem/762/B
 #greedy #two-pointer #sorting
*/
import java.util.Scanner;
import java.util.TreeSet;
import java.util.ArrayList;
import java.util.Collections;

public class UsbAndPS {
    public static boolean helper( int idx1, int idx2, int[] inputArr) {
        if (inputArr[idx1] > 0) {
            inputArr[idx1] -= 1;
            return true;
        } else if (inputArr[idx2] > 0) {
            inputArr[idx2] -= 1;
            return true;
        }
        return false;
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int[] inputArr = new int[3];
        for (int i=0; i < inputArr.length; i++) {
            int tmp = sc.nextInt();
            inputArr[i] = tmp;
        }
        int m = sc.nextInt();
        ArrayList<Long> usbMouseArr = new ArrayList<>();
        ArrayList<Long> psMouseArr = new ArrayList<>();
        
        int[] countArr = new int[2];
        for(int i=0; i < m; i++) {
            long cost = sc.nextLong();
            String name = sc.next();
            if (name.equals("USB")) {
                usbMouseArr.add(cost);
            } else {
                psMouseArr.add(cost);
            }
        }
        Collections.sort(usbMouseArr);
        Collections.sort(psMouseArr);
        long[] sumArr = new long[2]; // 0 - sum of mouse, 1 - sum of cost.
        int idxUsb = 0;
        int idxPS = 0;
        while (idxUsb < usbMouseArr.size() && idxPS < psMouseArr.size()) {
            if (inputArr[0] == 0 && inputArr[1] == 0 && inputArr[2] == 0 ) break;
           
            if (psMouseArr.get(idxPS) < usbMouseArr.get(idxUsb)) {
                if (helper(1, 2, inputArr)) {
                    sumArr[0] += 1;
                    sumArr[1] += psMouseArr.get(idxPS);
                    idxPS++;
                } else {
                    break;
                }
               
            } else  {
                if (helper(0, 2, inputArr)) {
                    sumArr[0] += 1;
                    sumArr[1] += usbMouseArr.get(idxUsb);
                    idxUsb++;
                } else {
                    break;
                }
            }
        }
        
        while (idxUsb < usbMouseArr.size()) {
            if (helper(0, 2, inputArr)) {
                sumArr[0] += 1;
                sumArr[1] += usbMouseArr.get(idxUsb);
                idxUsb++;
            } else {
                break;
            }
        } 
        
        while (idxPS < psMouseArr.size()) {
            if (helper(1, 2, inputArr)) {
                sumArr[0] += 1;
                sumArr[1] += psMouseArr.get(idxPS);
                idxPS++;
            } else {
                break;
            }
        }

        System.out.println(sumArr[0] +  " " + sumArr[1]);
    }
}

File: C++/UsbAndPS.cpp
/*  https://codeforces.com/problemset/problem/762/B
   #greedy #two-pointer #sorting
*/
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

bool helper(int& t1, int& t2) {
    if (t1 > 0) {
        t1--;
        return true;
    } else if (t2 > 0) {
        t2--;
        return true;
    }
    return false;
}

int main() {

    int a, b, c;
    cin >> a >> b >> c;
    int m;
    cin >> m;
    vector<int> usbMouseVector(m);
    vector<int> psMouseVector(m); 
    int countUSB=0, countPS = 0;
    int cost;
    string name;
    for(int i=0; i < m; i++) {
        cin >> cost >> name;
        if (name == "USB") {
            usbMouseVector[countUSB] = cost;
            countUSB++;
        } else {
            psMouseVector[countPS] = cost;
            countPS++;
        }
    }
    usbMouseVector.resize(countUSB);
    psMouseVector.resize(countPS);

    sort(usbMouseVector.begin(), usbMouseVector.end());
    sort(psMouseVector.begin(), psMouseVector.end());

    int sumComputer = 0;
    long long sumCost = 0;
   
    int idxUSB = 0;
    int idxPS = 0;

    while (idxPS < psMouseVector.size() && idxUSB < usbMouseVector.size()) {
        if ( a == 0 && b == 0 && c == 0 ) break;
        if ((psMouseVector[idxPS]) < usbMouseVector[idxUSB]) {
            if (helper(b, c)) {
                sumComputer += 1;
                sumCost += psMouseVector[idxPS];
                ++idxPS; 
            } else {
                break;
            }
        } else  {
            if (helper(a, c)){
                sumComputer += 1;
                sumCost += usbMouseVector[idxUSB]; 
                ++idxUSB;
            } else {
                break;
            }
        }
    }
    
    while (idxUSB < usbMouseVector.size()) {
        if (helper(a, c)){
            sumComputer += 1;
            sumCost += usbMouseVector[idxUSB]; 
            ++idxUSB;
        } else {
            break;
        }
    } 
    
    while (idxPS != psMouseVector.size() ) {
        if (helper(b, c)) {
            sumComputer += 1;
            sumCost += psMouseVector[idxPS];
            ++idxPS; 
        } else {
            break;
        }
       
    } 
    
    cout << sumComputer << " " << sumCost << endl;

    return 0;
}